Tabular Link Status Command

lsgunth opened the pull request:
This is the first crack at the tabular link status format. (3.1.4 in the work order).

Pretty straight forward but I could not figure out what the dl_active field in the table represented. Please let me know and I'll update the PR. 

Also, I made some basic assumptions about the port_active_lane field which should be reviewed.

I've also left the LTSSM strings as they are in the current code, so they print with parenthesis instead of a dash.

The example I was given is:
```
0x00000000:0000>lnkstat
[00] part:00.01 w:cfg[x04]-neg[x00] stk:0.0 port_active_lane:xxxx             dsp dl_active:0 Rate:G1 LTSSM: Detect -QUIET
[02] part:      w:cfg[x04]-neg[x00] stk:0.2 port_active_lane:xxxx             dsp dl_active:0 Rate:G1 LTSSM: Detect -QUIET
[04] part:00.02 w:cfg[x04]-neg[x00] stk:0.4 port_active_lane:xxxx             dsp dl_active:0 Rate:G1 LTSSM: Detect -QUIET
```
And the output on my switch is:
```
$ ./switchtec status -f table
[08] part:00.01 w:cfg[x08]-neg[x08] stk:1.0 port_active_lane:xxxxxxxx         dsp dl_active: 0 rate:G3 LTSSM: L0 (L0)
[12] part:00.02 w:cfg[x08]-neg[x08] stk:1.4 port_active_lane:xxxxxxxx         dsp dl_active: 0 rate:G3 LTSSM: L0 (L0)
[24] part:00.03 w:cfg[x16]-neg[x04] stk:3.0 port_active_lane:xxxxxxxxxxxxxxxx dsp dl_active: 0 rate:G3 LTSSM: L0 (L0)
[32] part:00.00 w:cfg[x16]-neg[x16] stk:4.0 port_active_lane:xxxxxxxxxxxxxxxx usp dl_active: 0 rate:G3 LTSSM: L0 (L0)
```

kelvin-cao commented on cli/main.c:
```diff
@@ -408,9 +459,13 @@ static int status(int argc, char **argv)
+		 .choices=output_fmt_choices},
 		{"reset", 'r', "", CFG_NONE, &cfg.reset_bytes, no_argument,
 		 "reset byte counters"},
 		{"verbose", 'v', "", CFG_NONE, &cfg.verbose, no_argument,
```
If the verbose mode only work for 'normal' output, better we have this info in the 'verbose' help.

kelvin-cao commented:
Normally our PRs target branch devel.

kelvin-cao commented:
It seems some white spaces are inserted between ':' and value in the output inconsistently, like 'dl_active: 0', while not in other places.

For the first_act_lane I think the valid values are 0 or 15 for a x16 lane in normal and reversal cases. @yungwesl please correct me if wrong.

lsgunth commented:
Thanks. Fixed up those issues. 

Definitely need a better understanding of the port_active_lane. It seems like it's a huge waste of space if it's just a number of x's based on the configured width.

yungwesl commented:
Hey Logan I’ll comment on GitHub on this. Port active lane is supposed to indicate not only which lanes are active but also what lane number they were assigned. When you are in l0 I am surprised you don’t see 01234567 or 76543210. The field is very useful is it displays several important details:

  *    The width
  *   Whether we negotiated a full width port or a sub port
  *   Which lanes in the port are active
  *   Which lane numbers were assigned
  *   The lane reversal status.


Cheers,
Wes
________________________________
From: Logan Gunthorpe ***@***.***>
Sent: Tuesday, April 6, 2021 9:05:57 AM
To: Microsemi/switchtec-user ***@***.***>
Cc: yungwesl ***@***.***>; Mention ***@***.***>
Subject: Re: [Microsemi/switchtec-user] Tabular Link Status Command (#258)

EXTERNAL EMAIL: Do not click links or open attachments unless you know the content is safe

Thanks. Fixed up those issues.

Definitely need a better understanding of the port_active_lane. It seems like it's a huge waste of space if it's just a number of x's based on the configured width.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/Microsemi/switchtec-user/pull/258#issuecomment-814241797>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AFZ3FXZRKMJ6L42EUGSL45DTHMWOLANCNFSM42NWHNVA>.

lsgunth commented:
Ah, that makes more sense. I was only going by the example you sent in the RFQ you sent me which only has xs.

I'll try to get the same info for my switch over the UART, but I could probably use a couple more examples.

Can you also describe what dl_active is supposed to mean?

lsgunth commented:
Hmm, lnkstat on my old switch does not show the part_active_lane field. This is what I see:

```
[08] part:00.01 w:cfg[x08]-neg[x08] stk:1.0 dsp dl_active:1 Rate:G3 LTSSM: L0     -L0 
[12] part:00.02 w:cfg[x08]-neg[x08] stk:1.4 dsp dl_active:1 Rate:G3 LTSSM: L0     -L0 
[24] part:00.03 w:cfg[x16]-neg[x04] stk:3.0 dsp dl_active:1 Rate:G3 LTSSM: L0     -L0 
[32] part:00.00 w:cfg[x16]-neg[x16] stk:4.0 usp dl_active:1 Rate:G3 LTSSM: L0     -L0 
```

yungwesl commented:
Hi Logan,

port_active_lane is derived from the following:
(1) Negotiated Link Width
(2) First Active Lane
(3) Lane Reversal

See the device specification chapter on MRPC 0x1C (Link Status Retrieve)

The pseudocode should be (Kelvin correct me if I am wrong)
- Read the Negotiated Link width to find out how many lanes are in this port 
-- (e.g. a port might be configured for x16 but only 8 lanes are active, so negotiated width is 8)
- Use First Active Lane to determine which lane is the first for the port 
-- e.g. On a x16 port, only two lanes can be first active (0 or 15) <- this is the lane 0 below
-- e.g. On a x8 port that forms inside a x16 configured port the valid first active lanes are (0,7,15)

- Then read the lane reversal field which has several options
-- 0x0 Normal lane ordering (0123456789abcdef, 01234567, 0123, 01, 0)
-- 0x1 x16 Reversal (fedcba9876543210, 76543210, 3210, 10, 0)
-- 0x2 x2 Reversal (10)
-- 0x4 x4 Reversal (3210)
-- 0x8 x8 Reversal (76543210)

- Using negotiated width and lane reversal plus first active lane, the reported information can be built.

This is probably the trickiest part of the code.  

Here is another example from Trident
--
```
0x00000000:000d>lnkstat
[00] part:01.00 w:cfg[x16]-neg[x16] stk:0.0 port_active_lane:fedcba9876543210 usp dl_active:1 Rate:G4 LTSSM: L0     -L0          
[08] part:03.01 w:cfg[x16]-neg[x00] stk:1.0 port_active_lane:xxxxxxxxxxxxxxxx dsp dl_active:0 Rate:G1 LTSSM: Detect -QUIET 
[16] part:02.00 w:cfg[x16]-neg[x16] stk:2.0 port_active_lane:fedcba9876543210 usp dl_active:1 Rate:G4 LTSSM: L0     -L0          
[24] part:00.01 w:cfg[x16]-neg[x00] stk:3.0 port_active_lane:xxxxxxxxxxxxxxxx dsp dl_active:0 Rate:G1 LTSSM: Detect -QUIET 
[32] part:00.05 w:cfg[x16]-neg[x16] stk:4.0 port_active_lane:0123456789abcdef dsp dl_active:1 Rate:G4 LTSSM: L0     -L0          
[40] part:00.00 w:cfg[x16]-neg[x16] stk:5.0 port_active_lane:0123456789abcdef usp dl_active:1 Rate:G4 LTSSM: L0     -L0          
[48] part:03.00 w:cfg[x04]-neg[x04] stk:6.0 port_active_lane:0123             usp dl_active:1 Rate:G4 LTSSM: L0     -L0     
```

yungwesl commented:
> Can you also describe what dl_active is supposed to mean?
dl_active is a PCIe term indicate that the data link layer is active.  In example above you can see that Rate is Gen4, LTSSM is L0 and dl_active = 1.  In cases where the LTSSM is not L0, dl_active will be 0. 

Thinking about this more I think the CLI tool is trying to express the MRPC return field named "Link Up".  To be honest, the more logical thing to do is rename dl_active.
Suggest:
link:
-  and then put "up" when the value is 1 and "dn" when the link is down.  

While we are at it, let's simplify the term port_active_lane. That's a mouthful.  
Suggest: 
- lanes:

lsgunth commented:
Perfect thanks, that's all the information I need. I probably won't get to these changes until next week though.

lsgunth commented:
Ok, here's another shot at this. There's still some uncertainty about the lanes string, but I've done what makes the best sense I can make out of it.

My switch seems to only return 0 for the first active lane (even though some ports are reversed) and I'm not sure if that's normal or just because it's a bit old (but it needs to be handled sensibly anyway). 

I've ran the code comprehensively for a large number of possible inputs and attached the result. Some inputs may not actually be valid in the real world, but that's fine as it just means the code will never see them.

[comprehensive_lanes.txt](https://github.com/Microsemi/switchtec-user/files/6307157/comprehensive_lanes.txt)

Let me know if I got anything wrong here.

yungwesl commented:
> Ok, here's another shot at this. There's still some uncertainty about the lanes string, but I've done what makes the best sense I can make out of it.
> 
> My switch seems to only return 0 for the first active lane (even though some ports are reversed) and I'm not sure if that's normal or just because it's a bit old (but it needs to be handled sensibly anyway).
> 
> I've ran the code comprehensively for a large number of possible inputs and attached the result. Some inputs may not actually be valid in the real world, but that's fine as it just means the code will never see them.
> 
> [comprehensive_lanes.txt](https://github.com/Microsemi/switchtec-user/files/6307157/comprehensive_lanes.txt)
> 
> Let me know if I got anything wrong here.

yungwesl commented:
> [comprehensive_lanes.txt](https://github.com/Microsemi/switchtec-user/files/6307157/comprehensive_lanes.txt)
> 
> Let me know if I got anything wrong here.

Thanks Logan,

I neglected to point you to another table in the datasheet (sorry) which should have helped clarify the valid combinations.  I marked up your sample and attached it here.

[comprehensive_lanes_wy.txt](https://github.com/Microsemi/switchtec-user/files/6307379/comprehensive_lanes_wy.txt)

Let me know if you have any questions.

lsgunth commented:
As I mentioned, the not-possible lines do not matter. If the switch never provides those parameters it will never be seen by the user. This is just showing what the function will produce given certain inputs.

However, some of the not valid and duplicate lines are ones a I see with my old switch and I needed to handle sanely to prevent it from crashing. If newer switches never produce that result, then that is fine.

So can you point out any lines that are incorrect for a set of parameters the switch is known to report? I don't see any of those cases in your mark up, but there is a lot of noise.

yungwesl commented:
> So can you point out any lines that are incorrect for a set of parameters the switch is known to report? I don't see any of those cases in your mark up, but there is a lot of noise.

Got it.  I re-reviewed the list and there are no outright errors in the reporting of the valid states. As you said we can ignore the invalid states since they should not happen.  I do have one concern around the invalid states which is not related to your code (that is the value of the REV field).  Is there room to report the REV field in your output? This will tell me if we see a state and it doesn't make sense I have one more useful piece of information. 

Otherwise it looks good.  

Wes

lsgunth commented:
Whether there's room or not is up to you really.... The only related limit to line length is a standard 80 character terminal which we are already longer than so it's not really a concern.

I've pushed a change to add the rev. Let me know if you change your mind or want it moved to a different location. Here's an example:

```
[08] part:00.01 w:cfg[x08]-neg[x08] stk:1.0 lanes:01234567         rev:0 dsp link:1 rate:G3 LTSSM: L0 (L0)
[12] part:00.02 w:cfg[x08]-neg[x08] stk:1.4 lanes:01234567         rev:0 dsp link:1 rate:G3 LTSSM: L0 (L0)
[24] part:00.03 w:cfg[x16]-neg[x04] stk:3.0 lanes:3210xxxxxxxxxxxx rev:1 dsp link:1 rate:G3 LTSSM: L0 (L0)
[32] part:00.00 w:cfg[x16]-neg[x16] stk:4.0 lanes:fedcba9876543210 rev:1 usp link:1 rate:G3 LTSSM: L0 (L0)
```

yungwesl commented:
Looks good to me. I have no further comments.

lsgunth commented:
Ok, well we're going to need a proper github review. 

And what's the policy for merges? Do you want me to merge my own PRs or will someone else handle that?

yungwesl commented:
The process we discussed internally was for Kelvin’s team to support some unit testing and review of the code. 

This PR would be filed against the devel branch and then Kelvin would drive it from there. 

Kelvin is this aligned with your understanding? If not we can discuss offline and then get Logan more clear instructions.

kelvin-cao commented:
> 
> 
> Whether there's room or not is up to you really.... The only related limit to line length is a standard 80 character terminal which we are already longer than so it's not really a concern.
> 
> I've pushed a change to add the rev. Let me know if you change your mind or want it moved to a different location. Here's an example:
> 
> ```
> [08] part:00.01 w:cfg[x08]-neg[x08] stk:1.0 lanes:01234567         rev:0 dsp link:1 rate:G3 LTSSM: L0 (L0)
> [12] part:00.02 w:cfg[x08]-neg[x08] stk:1.4 lanes:01234567         rev:0 dsp link:1 rate:G3 LTSSM: L0 (L0)
> [24] part:00.03 w:cfg[x16]-neg[x04] stk:3.0 lanes:3210xxxxxxxxxxxx rev:1 dsp link:1 rate:G3 LTSSM: L0 (L0)
> [32] part:00.00 w:cfg[x16]-neg[x16] stk:4.0 lanes:fedcba9876543210 rev:1 usp link:1 rate:G3 LTSSM: L0 (L0)
> ```

All looks good to me, except one unexpected white space between 'LTSSM:' and value string.

kelvin-cao commented:
> 
> 
> The process we discussed internally was for Kelvin’s team to support some unit testing and review of the code.
> 
> This PR would be filed against the devel branch and then Kelvin would drive it from there.
> 
> Kelvin is this aligned with your understanding? If not we can discuss offline and then get Logan more clear instructions.

Yes, it's aligned. 

Some MRPC commands has been updated in Gen4, and I'm assuming Logan is using Gen3 device for the development, therefore there might be chances that the newly developed features work on Gen3 but fail on Gen4. 

I think we can go like Logan still work on Gen3 and create PR, I (we) review the code, unit test on Gen4, I'll merge it into devel if the unit test on Gen4 passed or I'll get the failure info back to Logan for code update. Logan could you provide the unit test you want me to do on Gen4 in each PR then we can re-play it on our Gen4 platform?

lsgunth commented:
Re-pushed to fix the whitespace issue.

Yes, I'll have a gen4 switch soon to test with.

The tests for this PR should be pretty simple. Run

switchtec status -f table
and
switchtec status

Then verify the data matches what is expected. The main thing is the lanes field. Testing with a few different switch configurations would help too.

kelvin-cao commented:
That'll be more efficient if you will have a Gen4 to test with. I was a bit worrying about that my test might be the bottle neck due to my bandwidth constraints. 

I'm working remotely today and I just found the serial port is disconnected, I'll see what I can do tomorrow in lab.

kelvin-cao commented:
I ran the commands on two Gen4 switches on my platform. Please check the log here: [Command status on Gen4.txt](https://github.com/Microsemi/switchtec-user/files/6363134/Command.status.on.Gen4.txt)

It seems that the lanes look good but the port reported in the table format is less than what it was reported either in the normal format or in the UART 'lnkstat'.

lsgunth commented:
Hmm. Good catch. This is caused by SWITCHTEC_MAX_PORTS being 48.... And some of the physical port IDs in your switch exceed 48.

So, the question is whether SWITCHTEC_MAX_PORTS should be higher.... Or it looks like physical ports are always even and we should have 

```
#define SWITCHTEC_MAX_PHYS_PORT_ID  (SWITCHTEC_MAX_PORTS * 2)
```

Thoughts?

yungwesl commented:
Oh good catch indeed.

Logan, there are some levels of complexity here but I want Kelvin to confirm what I am saying

(1) Saratoga (Gen3) - Total of 96 lanes max, where Ports represent 2 lanes counting from 0 to 47
(2) Trident (Gen4) - Total of 100 lanes max, where Ports 0 to 47 represent lane-pairs from 0 to 95 and then there are 4 additional ports (48,49,50,51) mapped 1:1 with lanes 96 to 99
(3) Harpoon (Gen5) - Total 100 lanes max, where ports 0 to 47 represent lane-pairs from 0 to 95 and there are 4 additional ports (56,57,58,59) mapped 1:1 with lanes 96 to 99

lsgunth commented:
So I'm hearing that SWITCHTEC_MAX_PORTS should be bumped up to 60... Does this make sense to you Kelvin?

kelvin-cao commented:
Wes, yes it's also my understanding. 

Logan, it makes sense to me.

lsgunth commented:
Force pushed. Rebased onto the latest devel branch and added a commit to increase SWITCHTEC_MAX_PORTS.

Kelvin can you try again with your system and ensure those ports are visible?

kelvin-cao commented:
Sure, they look pretty good.

```
(On switch 0)
0x00000000:0000>lnkstat
[00] part:00.01 w:cfg[x08]-neg[x00] stk:0.0 port_active_lane:xxxxxxxx         dsp dl_active:0 Rate:G1 LTSSM: Detect -QUIET
[04] part:00.02 w:cfg[x08]-neg[x00] stk:0.4 port_active_lane:xxxxxxxx         dsp dl_active:0 Rate:G1 LTSSM: Detect -QUIET
[08] part:00.03 w:cfg[x16]-neg[x08] stk:1.0 port_active_lane:xxxxxxxx76543210 dsp dl_active:1 Rate:G3 LTSSM: L0     -L0
[16] part:00.04 w:cfg[x08]-neg[x00] stk:2.0 port_active_lane:xxxxxxxx         dsp dl_active:0 Rate:G1 LTSSM: Detect -QUIET
[20] part:00.05 w:cfg[x08]-neg[x00] stk:2.4 port_active_lane:xxxxxxxx         dsp dl_active:0 Rate:G1 LTSSM: Detect -QUIET
[24] part:01.00 w:cfg[x04]-neg[x04] stk:3.0 port_active_lane:0123             usp dl_active:1 Rate:G3 LTSSM: L0     -L0
[26] part:02.00 w:cfg[x04]-neg[x00] stk:3.2 port_active_lane:xxxx             usp dl_active:0 Rate:G1 LTSSM: Detect -QUIET
[28] part:03.00 w:cfg[x04]-neg[x00] stk:3.4 port_active_lane:xxxx             usp dl_active:0 Rate:G1 LTSSM: Detect -QUIET
[30] part:04.00 w:cfg[x04]-neg[x00] stk:3.6 port_active_lane:xxxx             usp dl_active:0 Rate:G1 LTSSM: Detect -QUIET
[32] part:00.06 w:cfg[x04]-neg[x04] stk:4.0 port_active_lane:0123             dsp dl_active:1 Rate:G4 LTSSM: L0     -L0
[34] part:00.07 w:cfg[x04]-neg[x00] stk:4.2 port_active_lane:xxxx             dsp dl_active:0 Rate:G1 LTSSM: Detect -QUIET
[36] part:00.08 w:cfg[x04]-neg[x00] stk:4.4 port_active_lane:xxxx             dsp dl_active:0 Rate:G1 LTSSM: Detect -QUIET
[38] part:00.09 w:cfg[x04]-neg[x00] stk:4.6 port_active_lane:xxxx             dsp dl_active:0 Rate:G1 LTSSM: Detect -QUIET
[40] part:00.10 w:cfg[x16]-neg[x00] stk:5.0 port_active_lane:xxxxxxxxxxxxxxxx dsp dl_active:0 Rate:G1 LTSSM: Detect -QUIET
[48] part:00.11 w:cfg[x04]-neg[x00] stk:6.0 port_active_lane:xxxx             dsp dl_active:0 Rate:G1 LTSSM: Detect -QUIET
[50] part:      w:cfg[x04]-neg[x00] stk:6.2 port_active_lane:xxxx             dsp dl_active:0 Rate:G1 LTSSM: Detect -QUIET

$ sudo ./switchtec status -f table 0
[00] part:00.01 w:cfg[x08]-neg[x00] stk:0.0 lanes:xxxxxxxx         rev:0 dsp link:0 rate:G1 LTSSM:Detect (QUIET)
[04] part:00.02 w:cfg[x08]-neg[x00] stk:0.4 lanes:xxxxxxxx         rev:0 dsp link:0 rate:G1 LTSSM:Detect (QUIET)
[08] part:00.03 w:cfg[x16]-neg[x08] stk:1.0 lanes:xxxxxxxx76543210 rev:1 dsp link:1 rate:G3 LTSSM:L0 (L0)
[16] part:00.04 w:cfg[x08]-neg[x00] stk:2.0 lanes:xxxxxxxx         rev:0 dsp link:0 rate:G1 LTSSM:Detect (QUIET)
[20] part:00.05 w:cfg[x08]-neg[x00] stk:2.4 lanes:xxxxxxxx         rev:0 dsp link:0 rate:G1 LTSSM:Detect (QUIET)
[24] part:01.00 w:cfg[x04]-neg[x04] stk:3.0 lanes:0123             rev:0 usp link:1 rate:G3 LTSSM:L0 (L0)
[26] part:02.00 w:cfg[x04]-neg[x00] stk:3.2 lanes:xxxx             rev:0 usp link:0 rate:G1 LTSSM:Detect (QUIET)
[28] part:03.00 w:cfg[x04]-neg[x00] stk:3.4 lanes:xxxx             rev:0 usp link:0 rate:G1 LTSSM:Detect (QUIET)
[30] part:04.00 w:cfg[x04]-neg[x00] stk:3.6 lanes:xxxx             rev:0 usp link:0 rate:G1 LTSSM:Detect (QUIET)
[32] part:00.06 w:cfg[x04]-neg[x04] stk:4.0 lanes:0123             rev:0 dsp link:1 rate:G4 LTSSM:L0 (L0)
[34] part:00.07 w:cfg[x04]-neg[x00] stk:4.2 lanes:xxxx             rev:0 dsp link:0 rate:G1 LTSSM:Detect (QUIET)
[36] part:00.08 w:cfg[x04]-neg[x00] stk:4.4 lanes:xxxx             rev:0 dsp link:0 rate:G1 LTSSM:Detect (QUIET)
[38] part:00.09 w:cfg[x04]-neg[x00] stk:4.6 lanes:xxxx             rev:0 dsp link:0 rate:G1 LTSSM:Detect (QUIET)
[40] part:00.10 w:cfg[x16]-neg[x00] stk:5.0 lanes:xxxxxxxxxxxxxxxx rev:0 dsp link:0 rate:G1 LTSSM:Detect (QUIET)
[48] part:00.11 w:cfg[x04]-neg[x00] stk:6.0 lanes:xxxx             rev:0 dsp link:0 rate:G1 LTSSM:Detect (QUIET)
[50] part:      w:cfg[x04]-neg[x00] stk:6.2 lanes:xxxx             rev:0 dsp link:0 rate:G1 LTSSM:Detect (QUIET)


(On switch 1)
0x00000000:0000>lnkstat
[00] part:00.01 w:cfg[x08]-neg[x00] stk:0.0 port_active_lane:xxxxxxxx         dsp dl_active:0 Rate:G1 LTSSM: Detect -QUIET
[04] part:      w:cfg[x08]-neg[x00] stk:0.4 port_active_lane:xxxxxxxx         dsp dl_active:0 Rate:G1 LTSSM: Detect -QUIET
[08] part:00.02 w:cfg[x16]-neg[x00] stk:1.0 port_active_lane:xxxxxxxxxxxxxxxx dsp dl_active:0 Rate:G1 LTSSM: Detect -QUIET
[16] part:00.03 w:cfg[x08]-neg[x00] stk:2.0 port_active_lane:xxxxxxxx         dsp dl_active:0 Rate:G1 LTSSM: Detect -QUIET
[20] part:      w:cfg[x08]-neg[x00] stk:2.4 port_active_lane:xxxxxxxx         dsp dl_active:0 Rate:G1 LTSSM: Detect -QUIET
[24] part:00.04 w:cfg[x08]-neg[x00] stk:3.0 port_active_lane:xxxxxxxx         dsp dl_active:0 Rate:G1 LTSSM: Detect -QUIET
[28] part:      w:cfg[x08]-neg[x00] stk:3.4 port_active_lane:xxxxxxxx         dsp dl_active:0 Rate:G1 LTSSM: Detect -QUIET
[32] part:01.00 w:cfg[x16]-neg[x04] stk:4.0 port_active_lane:0123xxxxxxxxxxxx usp dl_active:1 Rate:G4 LTSSM: L0     -L0
[40] part:00.00 w:cfg[x16]-neg[x16] stk:5.0 port_active_lane:0123456789abcdef usp dl_active:1 Rate:G3 LTSSM: L0     -L0
[48] part:00.05 w:cfg[x04]-neg[x00] stk:6.0 port_active_lane:xxxx             dsp dl_active:0 Rate:G1 LTSSM: Detect -QUIET
[50] part:      w:cfg[x04]-neg[x00] stk:6.2 port_active_lane:xxxx             dsp dl_active:0 Rate:G1 LTSSM: Detect -QUIET

$ sudo ./switchtec status -f table 1
[00] part:00.01 w:cfg[x08]-neg[x00] stk:0.0 lanes:xxxxxxxx         rev:0 dsp link:0 rate:G1 LTSSM:Detect (QUIET)
[04] part:      w:cfg[x08]-neg[x00] stk:0.4 lanes:xxxxxxxx         rev:0 dsp link:0 rate:G1 LTSSM:Detect (QUIET)
[08] part:00.02 w:cfg[x16]-neg[x00] stk:1.0 lanes:xxxxxxxxxxxxxxxx rev:0 dsp link:0 rate:G1 LTSSM:Detect (QUIET)
[16] part:00.03 w:cfg[x08]-neg[x00] stk:2.0 lanes:xxxxxxxx         rev:0 dsp link:0 rate:G1 LTSSM:Detect (QUIET)
[20] part:      w:cfg[x08]-neg[x00] stk:2.4 lanes:xxxxxxxx         rev:0 dsp link:0 rate:G1 LTSSM:Detect (QUIET)
[24] part:00.04 w:cfg[x08]-neg[x00] stk:3.0 lanes:xxxxxxxx         rev:0 dsp link:0 rate:G1 LTSSM:Detect (QUIET)
[28] part:      w:cfg[x08]-neg[x00] stk:3.4 lanes:xxxxxxxx         rev:0 dsp link:0 rate:G1 LTSSM:Detect (QUIET)
[32] part:01.00 w:cfg[x16]-neg[x04] stk:4.0 lanes:0123xxxxxxxxxxxx rev:0 usp link:1 rate:G4 LTSSM:L0 (L0)
[40] part:00.00 w:cfg[x16]-neg[x16] stk:5.0 lanes:0123456789abcdef rev:0 usp link:1 rate:G3 LTSSM:L0 (L0)
[48] part:00.05 w:cfg[x04]-neg[x00] stk:6.0 lanes:xxxx             rev:0 dsp link:0 rate:G1 LTSSM:Detect (QUIET)
[50] part:      w:cfg[x04]-neg[x00] stk:6.2 lanes:xxxx             rev:0 dsp link:0 rate:G1 LTSSM:Detect (QUIET)

```

kelvin-cao commented:
Do we still have anything to do for this PR before approval?

lsgunth commented:
As far as I'm concerned, it is good to go.
